src/runtime_src/core/tools/common/XBUtilities.cpp

```diff
@@ -280,6 +280,24 @@ XBUtilities::xrt_version_cmp(bool isUserDomain)
   }
 }
 
+static std::shared_ptr<xrt_core::device>
+get_device_internal(int index, bool in_user_domain)
+{
+  if (in_user_domain) {
+    static std::unordered_map<xrt_core::device::id_type, std::shared_ptr<xrt_core::device>> user_device_map;
+    if (user_device_map.find(index) == user_device_map.end())
+      user_device_map[index] = xrt_core::get_userpf_device(index);
+
+    return user_device_map[index];
+  }
+
+  static std::unordered_map<xrt_core::device::id_type, std::shared_ptr<xrt_core::device>> mgmt_device_map;
+  if (mgmt_device_map.find(index) == mgmt_device_map.end())
+    mgmt_device_map[index] = xrt_core::get_mgmtpf_device(index);
+
+  return mgmt_device_map[index];
+}
+
```

Comment:
Is std::unordered_map a bit of an overkill here ?  And the function is not thread safe, is that okay, else protect with static mutex at first line followed by lock_guard.
How about
```suggestion
static std::shared_ptr<xrt_core::device>
get_device_internal(int index, bool in_user_domain)
{
  static std::mutex mutex;
  std::lock_guard lk(mutex);
  
  if (in_user_domain) {
    static std::vector<std::shared_ptr<xrt_core::device>> user_device_map { xrt_core::get_total_devices(true /*user/*) };
    auto& device = user_device_map.at(index);
    if (!device)
     device = xrt_core::get_userpf_device(index);

    return device;
  }

  static std::vector<std::shared_ptr<xrt_core::device>> mgmt_device_map { xrt_core::get_total_devices(false /*mgmt*/) };
  auto& device = mgmt_device_map.at(index);
  if (!device)
    device = xrt_core::get_mgmtpf_device(index));

  return device;
}

```